webhook/resources/backuptarget/validator.go

```diff
@@ -0,0 +1,221 @@
+	namespace, err := b.ds.GetLonghornNamespace()
+	if err != nil {
+		return errors.Wrapf(err, "failed to get Longhorn namespace")
+	}
+	secret, err := b.ds.GetSecretRO(namespace.Name, secretName)
+	if err != nil {
+		if !apierrors.IsNotFound(err) {
+			return errors.Wrapf(err, "failed to get the secret before modifying backup target credential secret")
+		}
+		return nil
+	}
+	checkKeyList := []string{
+		types.AWSIAMRoleAnnotation,
+		types.AWSIAMRoleArn,
+		types.AWSAccessKey,
+		types.AWSSecretKey,
+		types.AWSEndPoint,
+		types.AWSCert,
+		types.CIFSUsername,
+		types.CIFSPassword,
+		types.AZBlobAccountName,
+		types.AZBlobAccountKey,
+		types.AZBlobEndpoint,
+		types.AZBlobCert,
+		types.HTTPSProxy,
+		types.HTTPProxy,
+		types.NOProxy,
+		types.VirtualHostedStyle,
+	}
+	for _, checkKey := range checkKeyList {
+		if value, ok := secret.Data[checkKey]; ok {
+			if strings.TrimSpace(string(value)) != string(value) {
+				return fmt.Errorf("there is space or new line in %s", checkKey)
+			}
+		}
+	}
+	return nil
+}
+
+func (b *backupTargetValidator) validateDRVolume(backupTarget *longhorn.BackupTarget) error {
+	vs, err := b.ds.ListDRVolumesRO()
+	if err != nil {
+		return errors.Wrapf(err, "failed to list standby volume when modifying BackupTarget")
+	}
+
+	backupTargetURL, err := getBackupTargetURLWithoutOpts(backupTarget.Spec.BackupTargetURL)
+	if err != nil {
+		return err
+	}
+
+	standbyVolumeNames := sets.NewString()
+	for k, v := range vs {
+		backupTargetURLfromBackupURL, err := getBackupTargetURLWithoutOpts(v.Spec.FromBackup)
+		if err != nil {
+			return err
+		}
+		fromBackupTargetName, ok := v.Labels[types.LonghornLabelBackupTarget]
+		if (ok && fromBackupTargetName == backupTarget.Name) || backupTargetURLfromBackupURL == backupTargetURL {
+			standbyVolumeNames.Insert(k)
+		}
+	}
+	if standbyVolumeNames.Len() != 0 {
+		return fmt.Errorf("cannot modify BackupTarget since there are existing standby volumes: %v", standbyVolumeNames)
+	}
+
+	return nil
+}
+
+func getBackupTargetURLWithoutOpts(inputURL string) (string, error) {
+	parsedURL, err := url.Parse(inputURL)
+	if err != nil {
+		return "", errors.Wrapf(err, "failed to parse URL %v", inputURL)
+	}
+	parsedURL.RawQuery = ""
+	return parsedURL.String(), nil
+}
+
+func (b *backupTargetValidator) Delete(request *admission.Request, oldObj runtime.Object) error {
+	backupTarget := oldObj.(*longhorn.BackupTarget)
+
+	var exists bool
+	if backupTarget.Annotations == nil {
+		// Annotations are nil, so the key does not exist
+		exists = false
+	} else {
+		// Annotations `DeleteBackupTargetFromLonghorn` is used to delete default backup target by Longhorn during uninstalling.
+		_, exists = backupTarget.Annotations[types.GetLonghornLabelKey(types.DeleteBackupTargetFromLonghorn)]
+	}
+	if backupTarget.Name == types.DefaultBackupTargetName && !exists {
+		return werror.NewInvalidError("deleting default backup target is not allowed", "")
+	}
+
```

Comment:
`exits` flag is not used by each backupTarget CR, so need to refactor the codes


```suggestion
	if backupTarget.Name == types.DefaultBackupTargetName {
        exists := ......
        if !exists {
		    return werror.NewInvalidError("deleting default backup target is not allowed", "")
        }
	}
```

Reply:
Modified. Thanks